monoio-http/src/h2/proto/ping_pong.rs

```diff
@@ -215,20 +211,15 @@ impl ReceivedPing {
 
 impl UserPings {
     pub(crate) fn send_ping(&self) -> Result<(), Option<proto::Error>> {
-        let prev = self
-            .0
-            .state
-            .compare_exchange(
-                USER_STATE_EMPTY,        // current
-                USER_STATE_PENDING_PING, // new
-                Ordering::AcqRel,
-                Ordering::Acquire,
-            )
-            .unwrap_or_else(|v| v);
+        let user_pings = unsafe { &mut *self.0.get() };
+        let prev = user_pings.state; // current
+        if prev == USER_STATE_EMPTY {
+            user_pings.state = USER_STATE_PENDING_PING; // new
+        }
 
         match prev {
             USER_STATE_EMPTY => {
-                self.0.ping_task.wake();
+                user_pings.ping_task.take().unwrap().wake();
```

Comment:
Here we should create a new `waker.rs` like what atomic-waker does.
1. Register waker should compare waker like `old_waker.will_wake(waker)`, if these two wakers not equal, then do waker.clone(), replace and wake old waker.
2. `wake` should not panic when inner waker is None. Use if let.

Reply:
fixed